src/category_theory/structured_arrow.lean

```diff
@@ -0,0 +1,193 @@
+/-
+Copyright (c) 2021 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Adam Topaz, Scott Morrison
+-/
+import category_theory.comma
+import category_theory.punit
+import category_theory.limits.shapes.terminal
+
+/-!
+# The category of "structured arrows"
+
+For `T : C ⥤ D`, a `T`-structured arrow with source `S : D`
+is just a morphism `S ⟶ T.obj Y`, for some `Y : D`.
+
+These form a category with morphisms `g : Y ⟶ Y'` making the obvious diagram commute.
+
+We prove that `𝟙 (T.obj Y)` is the initial object in `T`-structured objects with source `T.obj Y`.
+-/
+
+namespace category_theory
+
+universes v₁ v₂ u₁ u₂ -- morphism levels before object levels. See note [category_theory universes].
+variables {C : Type u₁} [category.{v₁} C] {D : Type u₂} [category.{v₂} D]
+
+/--
+The category of `T`-structured arrows with domain `S : D` (here `T : C ⥤ D`),
+has as its objects `D`-morphisms of the form `S ⟶ T Y`, for some `Y : C`,
+and morphisms `C`-morphisms `Y ⟶ Y'` making the obvious triangle commute.
+-/
+@[derive category, nolint has_inhabited_instance]
+def structured_arrow (S : D) (T : C ⥤ D) := comma (functor.from_punit S) T
+
+namespace structured_arrow
+
+variables {S S' S'' : D} {Y Y' : C} {T : C ⥤ D}
+
+/-- Construct a structured arrow from a morphism. -/
+def mk (f : S ⟶ T.obj Y) : structured_arrow S T := ⟨⟨⟩, Y, f⟩
+
+@[simp] lemma mk_left (f : S ⟶ T.obj Y) : (mk f).left = punit.star := rfl
+@[simp] lemma mk_right (f : S ⟶ T.obj Y) : (mk f).right = Y := rfl
+@[simp] lemma mk_hom_eq_self (f : S ⟶ T.obj Y) : (mk f).hom = f := rfl
+
+lemma eq_mk (f : structured_arrow S T) : f = mk (f.hom) :=
```

Comment:
```suggestion
lemma eq_mk (f : structured_arrow S T) : f = mk f.hom :=
```
Do you think this should be a simp lemma in the opposite direction?

Reply:
I think this direction is more useful because the other simp lemmas in this file simplify things built out of `mk f`.